### PR TITLE
Implement some of WeaponTemplate

### PR DESCRIPTION
**Pull request by jonwil**

### CODE REVIEW

File: src/game/logic/object/weapon.cpp [hunk 1]
```diff
@@ -457,3 +461,495 @@ bool Weapon::Is_Source_Object_With_Goal_Position_Within_Attack_Range(
...
+    m_primaryDamageRadius(0.0f),
+    m_secondaryDamage(0.0f),
+    m_secondaryDamageRadius(0.0f),
+    m_attackRange(0.0f),
+    m_minimumAttackRange(0.0f),
+    m_requestAssistRange(0.0f),
+    m_aimDelta(0.0f),
+    m_scatterRadius(0.0f),
+    m_scatterTargetScalar(0.0f),
+    m_shockWaveAmount(0.0f),
```
Comment (xezon) [inline src/game/logic/object/weapon.cpp]: m_shockWave* have wrong position in initializer list

File: src/game/logic/object/weapon.cpp [hunk 2]
```diff
@@ -457,3 +461,495 @@ bool Weapon::Is_Source_Object_With_Goal_Position_Within_Attack_Range(
...
+    m_shockWaveRadius(0.0f),
+    m_shockWaveTaperOff(0.0f),
+    m_damageType(DAMAGE_EXPLOSION),
+    m_deathType(DEATH_NORMAL),
+    m_weaponSpeed(999999.0f),
+    m_minWeaponSpeed(999999.0f),
```
Comment (xezon) [inline src/game/logic/object/weapon.cpp:699]: `bool m_scaleWeaponSpeed` is not initialized.

File: src/game/logic/object/weapon.cpp [hunk 3]
```diff
@@ -457,3 +461,495 @@ bool Weapon::Is_Source_Object_With_Goal_Position_Within_Attack_Range(
...
+    m_weaponRecoil(0.0f),
+    m_minTargetPitch(DEG_TO_RADF(-180.0f)),
+    m_maxTargetPitch(DEG_TO_RADF(180.0f)),
+    m_radiusDamageAngle(DEG_TO_RADF(180.0f)),
+    m_projectileTemplate(nullptr),
+    m_damageDealtAtSelfPosition(false),
```
Comment (xezon) [inline src/game/logic/object/weapon.cpp]: This and various other variables below have wrong position in initializer list

File: src/game/logic/object/weapon.cpp [hunk 4]
```diff
@@ -457,3 +461,495 @@ bool Weapon::Is_Source_Object_With_Goal_Position_Within_Attack_Range(
...
+
+void WeaponTemplate::Parse_Scatter_Target(INI *ini, void *formal, void *store, const void *user_data)
+{
+    WeaponTemplate *tmplate = static_cast<WeaponTemplate *>(formal);
+    Coord2D target;
+    INI::Parse_Coord2D(ini, nullptr, &target, nullptr);
+    tmplate->m_scatterTarget.push_back(target);
+}
+
+const float _DURATION_MULT = 0.029999999f;
```
Comment (xezon) [inline src/game/logic/object/weapon.cpp]: 0.03f
https://www.h-schmidt.net/FloatConverter/IEEE754.html

File: src/game/logic/object/weapon.cpp [hunk 5]
```diff
@@ -457,3 +461,495 @@ bool Weapon::Is_Source_Object_With_Goal_Position_Within_Attack_Range(
...
+void WeaponTemplate::Parse_Shot_Delay(INI *ini, void *formal, void *store, const void *user_data)
+{
+    WeaponTemplate *tmplate = static_cast<WeaponTemplate *>(formal);
+    const char *str = ini->Get_Next_Token_Or_Null(ini->Get_Seps_Colon());
+
+    if (!strcasecmp(str, "Min")) {
+        tmplate->m_minDelayBetweenShots = INI::Scan_Int(ini->Get_Next_Token(ini->Get_Seps_Colon()));
+        const char *str2 = ini->Get_Next_Token_Or_Null(ini->Get_Seps_Colon());
+
+        if (!strcasecmp(str2, "Max")) {
```
Comment (xezon) [inline src/game/logic/object/weapon.cpp]: Style: strcasecmp() == 0

File: src/game/logic/object/weapon.cpp [hunk 6]
```diff
@@ -457,3 +461,495 @@ bool Weapon::Is_Source_Object_With_Goal_Position_Within_Attack_Range(
...
+
+        for (int i = 0; i < VETERANCY_COUNT; i++) {
+            if (m_fireOCLName[i].Is_Empty()) {
+                m_fireOCL[i] = nullptr;
+            } else {
+                m_fireOCL[i] = g_theObjectCreationListStore->Find_Object_Creation_List(m_fireOCLName[i].Str());
+                captainslog_dbgassert(m_fireOCL[i] != nullptr, "OCL %s not found in a weapon!", m_fireOCLName[i].Str());
+            }
+
+            if (m_fireOCLName[i].Is_Empty()) {
```
Comment (xezon) [inline src/game/logic/object/weapon.cpp]: This code block here looks like a duplicate of the above. Is this intended?

Perhaps was meant to be `m_projectileDetonationOCLName` ?

File: src/game/logic/object/weapon.cpp [hunk 7]
```diff
@@ -457,3 +461,495 @@ bool Weapon::Is_Source_Object_With_Goal_Position_Within_Attack_Range(
...
+}
+
+const float _DURATION_MULT = 0.029999999f;
+
+void WeaponTemplate::Parse_Shot_Delay(INI *ini, void *formal, void *store, const void *user_data)
+{
+    WeaponTemplate *tmplate = static_cast<WeaponTemplate *>(formal);
+    const char *str = ini->Get_Next_Token_Or_Null(ini->Get_Seps_Colon());
+
+    if (!strcasecmp(str, "Min")) {
```
Comment (xezon) [inline src/game/logic/object/weapon.cpp]: Style: strcasecmp() == 0

File: src/game/logic/object/weapon.cpp [hunk 8]
```diff
@@ -457,3 +461,495 @@ bool Weapon::Is_Source_Object_With_Goal_Position_Within_Attack_Range(
...
+    "WALLS",
+    "SMALL_MISSILES",
+    "BALLISTIC_MISSILES",
+    "CONTROLLED_STRUCTURES",
+    nullptr,
+};
+
+void Parse_All_Vet_Levels_ASCII_String(INI *ini, void *formal, void *store, const void *user_data)
+{
+    auto OCL_str = ini->Get_Next_Ascii_String();
```
Comment (xezon) [inline src/game/logic/object/weapon.cpp]: Perhaps can curb the use of Auto in these new INI functions.

File: src/game/logic/object/weapon.cpp [hunk 9]
```diff
@@ -457,3 +461,495 @@ bool Weapon::Is_Source_Object_With_Goal_Position_Within_Attack_Range(
...
+     { "WeaponSpeed", &INI::Parse_Velocity_Real, nullptr, offsetof(WeaponTemplate, m_weaponSpeed) },
+     { "MinWeaponSpeed", &INI::Parse_Velocity_Real, nullptr, offsetof(WeaponTemplate, m_minWeaponSpeed) },
+     { "ScaleWeaponSpeed", &INI::Parse_Bool, nullptr, offsetof(WeaponTemplate, m_scaleWeaponSpeed) },
+     { "WeaponRecoil", &INI::Parse_Angle_Real, nullptr, offsetof(WeaponTemplate, m_weaponRecoil) },
+     { "MinTargetPitch", &INI::Parse_Angle_Real, nullptr, offsetof(WeaponTemplate, m_minTargetPitch) },
+     { "MaxTargetPitch", &INI::Parse_Angle_Real, nullptr, offsetof(WeaponTemplate, m_maxTargetPitch) },
+     { "RadiusDamageAngle", &INI::Parse_Angle_Real, nullptr, offsetof(WeaponTemplate, m_radiusDamageAngle) },
+     { "ProjectileObject", &INI::Parse_AsciiString, nullptr, offsetof(WeaponTemplate, m_projectileObject) },
+     { "FireOCL", &Parse_All_Vet_Levels_ASCII_String, nullptr, offsetof(WeaponTemplate, m_fireOCL) },
+     { "VeterancyFireOCL", &Parse_Per_Vet_Level_ASCII_String, nullptr, offsetof(WeaponTemplate, m_fireOCL) },
```
Comment (xezon) [inline src/game/logic/object/weapon.cpp:627]: `m_fireOCL` is correct here for "VeterancyFireOCL" ?

Same for several other Veterancy* names below.

File: src/game/logic/object/weapon.cpp (hunk 1 above)
Comment (jonwil) [reply src/game/logic/object/weapon.cpp]: fixed.

File: src/game/logic/object/weapon.cpp (hunk 2 above)
Comment (jonwil) [reply src/game/logic/object/weapon.cpp:699]: fixed.

File: src/game/logic/object/weapon.cpp (hunk 3 above)
Comment (jonwil) [reply src/game/logic/object/weapon.cpp]: fixed.

File: src/game/logic/object/weapon.cpp (hunk 4 above)
Comment (jonwil) [reply src/game/logic/object/weapon.cpp]: fixed.

File: src/game/logic/object/weapon.cpp (hunk 6 above)
Comment (jonwil) [reply src/game/logic/object/weapon.cpp]: Fixed.

File: src/game/logic/object/weapon.cpp (hunk 7 above)
Comment (jonwil) [reply src/game/logic/object/weapon.cpp]: fixed.

File: src/game/logic/object/weapon.cpp (hunk 8 above)
Comment (jonwil) [reply src/game/logic/object/weapon.cpp]: fixed.

File: src/game/logic/object/weapon.cpp (hunk 9 above)
Comment (jonwil) [reply src/game/logic/object/weapon.cpp:627]: This is correct.

File: src/game/logic/object/weapon.cpp (hunk 5 above)
Comment (jonwil) [reply src/game/logic/object/weapon.cpp]: fixed.

**Comment by xezon** [comment]
Tested on Release. Shell Map looks as usual. Test map with practice targets looks as usual. No obvious issues found.

**Review by xezon — APPROVED** [review]

**Pull request merged**